### ISSUE TITLE
Constant evaluation cannot have code coverage instrumentation

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust#124732]
This is a hard problem to solve, but I figured I'd open an issue for this since it's an interesting problem that could potentially have a few solutions.

One "easy" solution would be to have a compiler flag to actually generate the code to evaluate constants as a separate test (or, potentially one test per constant) to be used by test harnesses, and then any existing instrumentation for tests could be applied to this. It would effectively be converting the below code:

```rust
const X: T = f();
const Y: U = g();
```

into the below code:

```rust
#[test]
fn eval_X() {
    let X: T = f();
    black_box(X);
}

#[test]
fn eval_Y() {
    let Y: T = g();
    black_box(Y);
}
```

Although it would require some changes to still apply intrinsics like `const_eval_select` properly as if the code were running in constant time. It would be still impossible to ensure that generated code runs exactly like the constant evaluation would, but it would certainly be "close enough" for ensuring the basics like if certain methods and branches have been called in constant code.

The main reason this occurred to me at all was because I actually noticed this difference when writing a few tests of my own: if I factor constructing a value into a constant instead of doing it each time in a test, I actually lose coverage for the constructor since it's never called at runtime. This is fine if the constructor is trivial, but if it contains logic for correcting certain kinds of errors, it would be nice to see those cases covered without having to duplicate the code and create my own separate test for it.

**Comment by RalfJung** [comment]
Could you update the issue description to start by explaining the problem? Currently it starts by saying "the problem is hard" without saying what the problem is. I was confused for a while here. Of course there's no coverage instrumentation, it's compile-time code after all...

**Comment by clarfonthey** [comment]
Right, I was avoiding providing explicit code examples because I felt that it might get a bit too in the weeds, but I'll try and add one to make it a bit easier to understand.

**Comment by clarfonthey** [comment]
Updated the description with hopefully a better example. Keep in mind that as I said, I don't think that this feature is worth having *now* given my current example, but as more stuff gets stabilised for const evaluation, it feels like this would be more and more desired.

**Comment by RalfJung** [comment]
`const fn` can be called at compiletime and runtime, so I am not entirely sure why there would be code duplication.

> To test the normalization code, I decided to create a bunch of sample spring states and compared their values in the test, but because I wanted to share them across tests, I constructed them in constants. This caused the normalization code to be entirely absent from code coverage, since it happened at compile time. Since the normalization after various operations isn't the same as the normalization in the constructor, the tests for those also didn't affect the constructor's code coverage.

There is another way to share them: write functions of type `const fn f() -> T`. Then you can call these functions at runtime and will get coverage information.

Arguably that's still a work-around, but the overhead compared to using consts in your tests seems rather small.

**Comment by clarfonthey** [comment]
> `const fn` can be called at compiletime and runtime, so I am not entirely sure why there would be code duplication.
>
> There is another way to share them: write functions of type `const fn f() -> T`. Then you can call these functions at runtime and will get coverage information.
> 
> Arguably that's still a work-around, but the overhead compared to using consts in your tests seems rather small.

Oh, like I said, I completely agree that this would mostly be overkill with the current state of const evaluation. It mostly just prompted the idea for me since I know that const eval will become substantially more advanced in the future.

**Comment by RalfJung** [comment]
Yes it will, but the approach of "just" having your test call the `const fn` at runtime will still work. So it seems to me like if coverage matters, you can keep obtaining coverage that way.

**Comment by oli-obk** [comment]
if it's not too much complexity in the mir interpreter, could the interpreter collect the coverage data during evaluation and then have it exposed at the use runtime sites of the constant? As long as most of the complexity is in the coverage logic and not the interpreter, that seems maintainable to me.

**Comment by clarfonthey** [comment]
Coverage instrumentation in miri in general feels like something that would be useful to have, although I'm not sure if that's already possible or not. Since the compiled code already links back to the original source, I would assume that the MIR also has access to the original source as well for storing that info.
